File: pubsub/src/main/java/glitch/pubsub/object/json/WhisperMessage.java
package glitch.pubsub.object.json;

import com.google.common.collect.ImmutableSet;
import com.google.gson.annotations.SerializedName;
import glitch.api.objects.enums.UserType;
import glitch.api.objects.json.Badge;
import glitch.api.objects.json.interfaces.IDObject;
import glitch.pubsub.object.json.message.Emote;
import lombok.Data;

import java.awt.*;
import java.time.Instant;

@Data
public class WhisperMessage implements IDObject<Long> {
    private final Long id;

    @SerializedName("body")
    private final String message;

    @SerializedName("sent_ts")
    private final Instant createdAt;

    @SerializedName("from_id")
    private final Long senderId;

    private final Tags tags;

    private final Recipient recipient;

    @Data
    public class Tags {
        private final String login;
        private final String displayName;
        private final Color color;
        private final UserType userType;
        private final ImmutableSet<Emote> emotes;
        private final ImmutableSet<Badge> badges;
    }

    @Data
    public class Recipient implements IDObject<Long> {
        private final Long id;
        private final String username;
        private final String displayName;
        private final Color color;
        private final UserType userType;
        private final ImmutableSet<Badge> badges;
        private final String profileImage;
    }
}

File: pubsub/src/main/kotlin/glitch/pubsub/object/json/WhisperMessage.kt
package glitch.pubsub.`object`.json

import com.google.gson.annotations.SerializedName
import glitch.api.objects.enums.UserType
import glitch.api.objects.json.Badge
import glitch.api.objects.json.interfaces.IDObject
import glitch.pubsub.`object`.json.message.Emote
import java.awt.Color
import java.time.Instant

class WhisperMessage(
        override val id: Long,
        @SerializedName("body")
        val message: String,
        @SerializedName("sent_ts")
        val createdAt: Instant,
        @SerializedName("from_id")
        val senderId: Long, val tags: Tags,
        val recipient: Recipient
) : IDObject<Long> {

    data class Tags(
            val login: String,
            val displayName: String,
            val color: Color?,
            val userType: UserType,
            val emotes: Set<Emote>,
            val badges: Set<Badge>
    )

    data class Recipient(
            override val id: Long,
            val username: String,
            val displayName: String,
            val color: Color?,
            val userType: UserType,
            val badges: Set<Badge>,
            val profileImage: String
    ) : IDObject<Long>
}
